src/shared/mount-util.c

```diff
@@ -1035,6 +1016,179 @@ static int mount_in_namespace(
+        if (r < 0)
+                return log_debug_errno(r, "Failed to resolve source path of %s: %m", src);
+        log_debug("Chased source path of %s to %s", src, chased_src_path);
+
+        if (fstat(chased_src_fd, &st) < 0)
+                return log_debug_errno(errno, "Failed to stat() resolved source path %s: %m", src);
+        if (S_ISLNK(st.st_mode)) /* This shouldn't really happen, given that we just chased the symlinks above, but let's better be safe… */
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Source directory %s can't be a symbolic link", src);
+
+        if (!mount_new_api_supported()) /* Fallback if we can't use the new mount API */
+                return mount_in_namespace_legacy(
+                                chased_src_path,
+                                chased_src_fd,
+                                &st,
+                                propagate_path,
+                                incoming_path,
+                                dest,
+                                pidns_fd,
+                                mntns_fd,
+                                root_fd,
+                                read_only,
+                                make_file_or_directory,
+                                options,
+                                image_policy,
+                                is_image);
+
+        if (is_image) {
+                r = verity_dissect_and_mount(
+                                chased_src_fd,
+                                chased_src_path,
+                                /* dest= */ NULL,
+                                options,
+                                image_policy,
+                                /* required_host_os_release_id= */ NULL,
+                                /* required_host_os_release_version_id= */ NULL,
+                                /* required_host_os_release_sysext_level= */ NULL,
+                                /* required_sysext_scope */ NULL,
+                                &img);
+                if (r < 0)
+                        return log_debug_errno(
+                                        r,
+                                        "Failed to dissect and mount image %s: %m",
+                                        chased_src_path);
+        } else {
+                new_mount_fd = open_tree(
+                                chased_src_fd,
+                                "",
+                                OPEN_TREE_CLONE|OPEN_TREE_CLOEXEC|AT_SYMLINK_NOFOLLOW|AT_EMPTY_PATH);
+                if (new_mount_fd < 0)
+                        return log_debug_errno(
+                                        errno,
+                                        "Failed to open mount point \"%s\": %m",
+                                        chased_src_path);
+        }
+
+        if (read_only && new_mount_fd >= 0) {
+                if (mount_setattr(new_mount_fd, "", AT_EMPTY_PATH,
```

Comment:
`mount_new_api_supported()` only checks `fsopen()` but relies on `open_tree()` and `mount_setattr()`. At least `mount_setattr()` came way later than the other. So `mount_new_api_supported()` should probably do upper bounds checking, i.e., check for `mount_setattr()` or check for all mount api system calls which is safer because it also captures seccomp profiles that block stuff or kernels with partial backports. Back then I did something like:

```c
bool can_use_mount_api(void)
{
	static int supported = -1;

	if (supported == -1) {
		__do_close int fd = -EBADF;

		fd = openat2(-EBADF, "", NULL, 0);
		if (fd > 0 || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		fd = fsmount(-EBADF, 0, 0);
		if (fd > 0 || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		fd = fsconfig(-EBADF, -EINVAL, NULL, NULL, 0);
		if (fd > 0 || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		fd = fsopen(NULL, 0);
		if (fd > 0 || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		fd = move_mount(-EBADF, NULL, -EBADF, NULL, 0);
		if (fd > 0 || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		fd = open_tree(-EBADF, NULL, 0);
		if (fd > 0 || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		supported = 1;
		TRACE("Kernel supports mount api");
	}

	return supported == 1;
}

bool can_use_bind_mounts(void)
{
	static int supported = -1;

	if (supported == -1) {
		int ret;

		if (!can_use_mount_api()) {
			supported = 0;
			return false;
		}

		ret = mount_setattr(-EBADF, NULL, 0, NULL, 0);
		if (!ret || errno == ENOSYS) {
			supported = 0;
			return false;
		}

		supported = 1;
		TRACE("Kernel supports bind mounts in the new mount api");
	}

	return supported == 1;
}
```

Reply:
ok I've switched to check for mount_setattr too